File: 프로그래머스/2/131127. 할인 행사/할인 행사.py
def solution(want, number, discount):
#   1가지 제품을 하루에 1개만 구매가능
#   정현이는 원하는 제품과 수량이 할인하는 날짜와 10일연속 일치시 회원가입
    
#   셋째 날부터..
    start = 0 
    end = 9
    
    cnt = 0
    for i in range(len(discount)):
        start = i
        end = start + 10
        
        if len(discount)-i < 10 :
            end = len(discount)
            
        
        dic = dict()
        for idx,item in enumerate(want):
            dic[item] = number[idx]
#       딕셔너리에 banana:3, apple:2 등록
        for j in range(start,end) :
            if discount[j] in dic :
                dic[discount[j]]-=1
                
        # print(dic)
        
        check = -1        
        for k,v in dic.items() :
            if v >= 1 :
                check = 0
        if check == -1 :
            cnt+=1
                
            
  
    answer = cnt
    return answer

File: 프로그래머스/2/131127. 할인 행사/할인 행사.java
import java.util.*;
class Solution {
    /**
    알고리즘 분류 : 해시 
    어떻게 풀이?
    1.첫날부터 number.size일 동안의 할인 목록을 구함
    2. 일치하면 +1 
    
    */
    public int solution(String[] want, int[] number, String[] discount) {
        int n = number.length;
        int days = discount.length;
        int answer = 0;
        Map<String,Integer> wantMap = new HashMap<>();
        
        for(int i=0; i<want.length; i++){
            wantMap.put(want[i],number[i]);
        }
        
        for(int i =0; i<days-9; i++){
            Map<String, Integer> curDiscount = new HashMap<>();
            
            //해당 일에 회원가입하면 10일동안의 할인 메뉴 모음
            for(int j=i; j<i+10; j++){
                if(!curDiscount.containsKey(discount[j])){
                    curDiscount.put(discount[j],1);
                }
                else{
                    curDiscount.put(discount[j],curDiscount.get(discount[j])+1);
                }
                    
            }
            
            if(wantMap.equals(curDiscount)){
                answer++;
            }
        }
        
        
        return answer;
    }
}